slides/src/content/rxjs-observables.md

```diff
@@ -1,27 +1,127 @@
+* Both `Promises` and `Observables` are used for dealing with asynchronous code. However, 
+Observables are considered more powerful because:
+  * `Observables` are more expressive.
+  * `Observables` are cancellable.
+  * `Observables` have rich API that make operations like retrying much easier (*i.e.*, with `retry` and 
+  `retryWhen` operators).
 
 ---
 
-## Search with `switchMap`
+## Creating an Observable
 
-The one downside to using `flatMap` for our search feature is that we will get results for all events in the stream and not necessarily in order.
-This would be a huge problem for our search feature as we may not get the latest results back.
+```js
+import {Observable} from 'rxjs/Observable';
+
+const observable = new Observable(observer => {
+  observer.next(1);
+  observer.next(2);
+  observer.complete();
+})
+```
+Creating the Observable alone does not trigger anything.
+Creating an Observable is like defining a function: 
+- A function does not do anything until it is *called*. 
+- Likewise, an Observable does not start streaming data until it's been *subscribed*.
+
+---
 
-`switchMap` solves this problem by only subscribing to the latest event. Any time a new event comes in, `switchMap` will automatically
-unsubscribe from the older event to make sure you get the latest results.
+## Subscribing to an Observable
 
-<img src="content/images/switch-map.png" alt="flat-map-diagram" width="40%" />
+```js
+const subscription = observable.subscribe(
+  value => console.log(value),    // handles next
+  error => console.log(error),    // handles error 
+  () => console.log('complete'),  // handles complete
+);
+```
 
-[View Example](http://plnkr.co/edit/FYLTcx?p=preview)
+Subscribing to the Observable we created earlier would print:
+
+```
+1
+2
+complete
+```
+
+Also calling `subscribe` returns a *Subscription* object which is used to *dispose* or *cancel*
+execution of an Observable.
 
 ---
+
+## Disposing (Cancelling) Execution
+
+Sometimes we want to cancel an `Observable`'s execution (*i.e.*, for an infinite sequence). We can
+achieve this by *unsubscribing* from the executing Observable.
+
+```js
+const subscription = observable.subscribe(
+  value => console.log(value),   
+  error => console.log(error),    
+  () => console.log('complete'),  
+); 
+
+subscription.unsubscribe(); // disposing observable execution
+```
+
+---
+
+## Observables from Other (Angular) Sources
+
+### Observable HTTP Events
+
+```js 
+http.get('http://jsonplaceholder.typicode.com/users/')
+  .mergeMap((data) => data.json())
+  .subscribe((data) => {
+    this.doctors.push(data);
+  });
+```
+
+### Observable Form Events 
+
+```js
+this.email = new FormControl();
+
+this.email.valueChanges
+  .map(n => n.split('').reverse().join(''))
+  .subscribe(value => this.data = value);
+```
+
+---
+
+## Observable Operations
+
+```js
+http.get('http://jsonplaceholder.typicode.com/users/')
+  .mergeMap((response) => response.json())
```

Comment:
@tsungting 

I've played with `map` and `mergeMap` a bit. [See this Plunkr](https://plnkr.co/edit/c3DUKg?p=preview).

It seems that `mergeMap` "converts" result of `data.json()` (which is a JavaScript array) into an Observable stream that emits each element one at a time so that `this.doctors.push(data)` would just work.

Whereas if we used `map`, subscription would get a single emitted `data` via creating a response meta-stream that emits data once as a whole array, which when subscribed, we would need to iterate over in order to add to `this.doctors` as such:
```
.map(data => data.json())
.subscribe((data) => {
      data.forEach(doc => this.doctors.push(doc.name))
      //this.doctors.push(data.name); can't do this anymore
          
      cd.detectChanges();
});
```

Also, regarding your point with Promise, I believe in Promises, you don't need to strictly return a new Promise in order to allow chaining:

````
Promise.then(function() {
    return 'Hello';
}).then(function(data) {
   console.log(data); // hello
})
```
would also work.